code/centauri/ibc/modules/README.md

```diff
@@ -1,47 +1,214 @@
 # IBC module
 
-[![Crate][crate-image]][crate-link]
-[![Docs][docs-image]][docs-link]
-[![Build Status][build-image]][build-link]
-[![End to End testing][e2e-image]][e2e-link]
 [![Apache 2.0 Licensed][license-image]][license-link]
-![Rust Stable][rustc-image]
-![Rust 1.51+][rustc-version]
 
+Implementation of the Inter-Blockchain Communication Protocol ([IBC]) in rust.
 
-See the [ibc-rs] repo root for more detailed information on how this crate can be used.
+## Project Structure
 
+**Core** contains the traits and handlers that enable the ibc protocol.  
+**Applications** contains sub protocols that are built off the core ibc.  
+**Mock** contains implementations of the core ibc protocol for testing purposes.  
 
-Implementation of the Inter-Blockchain Communication Protocol ([IBC]) module.
+## Architecture
 
-## Documentation
+The design of this crate is geared towards making it as performant as possible and reducing runtime overhead to the bear minimum.  
+The latter is a major focus because it is expected that this crate will be executed inside blockchain runtimes, which are extremely resource constrained environments  
+and anything that could possibly reduce code execution time needs to be prioritized. 
+To achieve these noble goals, this crate is designed to prioritize static dispatch over dynamic dispatch, while the former is less flexible from a development perspective, it helps could potentially help meet performance 
+requirements for blockchain runtimes. We also provide a couple procedural macros in the [`ibc-derive`](/code/centauri/ibc/derive) crate,  
+that abstracts away a lot of boilerplate code for the end user.
+
+The framework is mostly defined as a set of traits that need to be implemented to make use of the message handling capabilities.
+
+## Terminology
+
+A couple definitions to help understand the architecture of this framework
+- **Reader** - A `Reader` is a trait that defines methods that provide read access to the underlying storage of the host.
+- **Keeper** - A `Keeper` trait is one that defines methods that provide write access to the underlying storage of the host.
+- **Context** - The context is a type that implements all the Reader, Keeper and Routing traits that govern access to the storage of the host and module callbacks.
+- **Handler** - A handler is a function that handles processing of an ibc message type and returns a result.
+- **Event** - A struct which when emitted signifies successful processing of a message.
+- **Router** - A type that channels packets to the correct module for handling.
+
+## ICS02 Client Definitions
+
+The client module hosts the trait definitions, messages and handlers for light client implementation.
```

Comment:
hosts trait definitions and handlers clear. so `messages` not clear. how one may host message?